src/cargo/sources/git/utils.rs

```diff
@@ -461,24 +463,23 @@ impl<'a> GitCheckout<'a> {
                 }
             };
             // Fetch data from origin and reset to the head commit
-            let reference = GitReference::Rev(head.to_string());
             gctx.shell()
                 .status("Updating", format!("git submodule `{child_remote_url}`"))?;
-            fetch(
-                &mut repo,
-                &child_remote_url,
-                &reference,
+            let mut source = GitSource::new(
+                SourceId::from_url(&format!("git+{child_remote_url}#{head}"))?,
                 gctx,
-                RemoteKind::GitDependency,
             )
             .with_context(|| {
                 let name = child.name().unwrap_or("");
                 format!("failed to fetch submodule `{name}` from {child_remote_url}",)
             })?;
```

Comment:
Seems like we should have this also on `update_db` and `copy_to`

Reply:
See an alternative in <https://github.com/rust-lang/cargo/pull/16246#discussion_r2573126320>